### PR TITLE
Reward history

### LINKED ISSUE #1314: JSON reward export
**Is your feature request related to a problem/context ? Please describe if applicable.**
Please implement a json export format for rewards in addition to the existing CSV.

### LINKED ISSUE #1348: Rewards per epoch not fully credited
**Describe the bug**

There was one pool created on a private testnet during genesis block + 2 delegators delegating to it. Pool tax is 10%, reward per epoch - 1000. No treasury enabled. During epoch 8 was created 2 blocks, pool got 100 and two delegators 449 lovelaces each. Total distributed reward - 998 instead of 1000.

**Mandatory Information**
1. jcli 0.8.0 (HEAD-7868354, release, linux [x86_64]) - [rustc 1.39.0 (4560ea788 2019-11-04)]
2. jormungandr 0.8.0 (HEAD-7868354, release, linux [x86_64]) - [rustc 1.39.0 (4560ea788 2019-11-04)]

**To Reproduce**
1. Create new blockchain with provided genesis settings below.
2. Wait for any epoch to end. 
3. Check how many blocks were created during that epoch.
4. Calculate pool tax and remaining rewards that should be distributed.
5. Compare it with data returned by jcli.

**Expected behavior**
Total sum of distributed rewards should be 1000.

**Additional context**
genesis.yaml:
https://pastebin.com/4gW7AwER

Pool's owner rewards for epoch 8:
Screenshot 2019-12-12 at 12 32 22

Delegator1 rewards for epoch 8:
Screenshot 2019-12-12 at 12 34 02

Delegator2 rewards for epoch 8:
Screenshot 2019-12-12 at 12 35 21

Pool info (don't mind balances, jcli call was made on epoch 10):
Screenshot 2019-12-12 at 12 37 10

### PR DESCRIPTION
**Pull request by NicolasDP**
Add REST API to access the reward distributions (note that for a full reward distribution dump you need `--rewards-report-all` to be set when starting the node.

WARNING: using `--rewards-report-all` will have an impact on the node's CPU and memory. If you are on a small CPU machine or if you don't need to see all reward history for everyone: do not use. For example: It is not necessary for a wallet client or for a stake pool. On an explorer or a tool to monitor stake pool rewards or performance: it is useful and worth the memory/CPU consumption.

adds the following endpoints:

* `GET /api/v0/rewards/history/{length}`: which gives the reward history for the `length` last epoch;
* `GET /api/v0/rewards/epoch/{epoch}`: which gives the reward history for the given epoch `epoch`;

closes #1314 : rewards can be seen in JSON format... 
closes #1348 : it is possible to check the reward distribution of the whole epoch and so possible to validate that actually no rewards was due to be distributed.

---

_EDIT_: made the warning shot less strong.

### CODE REVIEW

File: jormungandr/src/rest/v0/handlers.rs [hunk 1]
```diff
@@ -389,6 +389,70 @@ pub async fn get_network_stats(context: Data<Context>) -> Result<impl Responder,
...
+        return Err(ErrorNotFound("Invalid epoch, does not exist yet..."));
+    }
+
+    loop {
+        if tip_ref.block_date().epoch == epoch {
+            break;
+        } else {
+            if let Some(previous_epoch) = tip_ref.last_ref_previous_epoch() {
+                assert_eq!(
+                    previous_epoch.block_date().epoch + 1,
```
Comment (NicolasDP) [inline jormungandr/src/rest/v0/handlers.rs]:
```suggestion
                    previous_epoch.block_date().epoch,
```

**Comment by darko-mijic** [comment]
> Do not use on a wallet client or stake pool

Why would this be merged?

**Comment by NicolasDP** [comment]
> > Do not use on a wallet client or stake pool
> 
> Why would this be merged?

Quoting few words of a sentence is not going to help your cause. What’s your concern? You think your team will enable this option without reading the documentation of this command line parameter?

**Comment by dorin100** [comment]
I tested the feature, working

**Pull request merged**